mustache.mjs

```diff
@@ -647,16 +660,38 @@ Writer.prototype.unescapedValue = function unescapedValue (token, context) {
     return value;
 };
 
-Writer.prototype.escapedValue = function escapedValue (token, context) {
+Writer.prototype.escapedValue = function escapedValue (token, context, config) {
+  var escape = this.getConfigEscape(config) || mustache.escape;
   var value = context.lookup(token[1]);
   if (value != null)
-    return typeof value === 'number' ? String(value) : mustache.escape(value);
+    return (typeof value === 'number' && escape === mustache.escape) ? String(value) : escape(value);
```

Comment:
Reading `&& escape === mustache.escape` gives me the impression we do not escape the value if the default escape function is being used?

Reply:
..I certainly would have expected a test or two to explode badly if the default HTML escaping got removed by this change.

Since that's not the case, I guess I'm basically wondering why I'm not really grasping why this introduced change is needed 😄

Reply:
This ternary condition was from a recent PR (https://github.com/janl/mustache.js/pull/754) that was intended to optimize performance when writing  numeric values. It works fine for escaping HTML, but someone might want their custom escape function to treat numbers differently from the default JS number stringification behavior. So now template rendering will continue using this optimized path when the escape function is the global default and skip it otherwise.

Perhaps one could make an argument for moving this optimization out of this function and into the default escape function? I'm fine making that change, I think it would certainly make the code more clearly legible if nothing else. It would also avoid a potential pitfall with modifying the global escape function and expecting to be able to write numbers differently.

Though I'd note that this pitfall is preserved behavior and not a newly introduced regression. Changing this behavior may break user code that was expecting their global custom escape function to never have to handle a number value.

Reply:
Since there are other changes that may break code using custom tags or escape functions that is sensitive to some of these quirks, I think the best thing would be to risk the breaking and move this check into the `escapeHtml` function. That function can just not go to the trouble of calling `String.replace` with a regex when the input is guaranteed to not contain HTML metacharacters, i.e. `null`, `undefined`, `true`, `false`, and numbers. Let me know what you think. If you agree, I'll go ahead and move that type test logic into `escapeHtml` instead.

Reply:
Here is what a more invasive change to the value escaping functionality might look like.

All tests continue to pass without changes, so this is _essentially_ backwards compatible, still. But it is likely to change behavior in ways that users relying on global escape function replacement might find inconvenient. If you like it I'll write some tests and make sure everything is solid. If not then the PR as it stands now still satisfies my own needs.

https://github.com/pineapplemachine/mustache.js/compare/render-with-provided-escape-function...pineapplemachine:render-config-invasive

Reply:
> It works fine for escaping HTML, but someone might want their custom escape function to treat numbers differently from the default JS number stringification behavior.

Aaaah very good point indeed, makes perfect sense.

I did not think of that aspect when reviewing #754 to be honest. The fact that we've made it so that non-default escape functions won't be able to escape `number`s as a result of that `typeof value === 'number'` check, is certainly not good. Thanks for pointing that out!

We'll have to reconsider that `number` check -- that can be done separate from this PR tho.

Reply:
I fully agree with the rationale behind introducing `&& escape === mustache.escape` that you shared above.

In danger of sounding too pedantic, I'm wondering if this change could be made in its own PR. Primarily to keep the history of changes as focused and "honest" as possible, without other seemingly unrelated changes introduced.

Also feel like what you've just brought up, is severe enough that it deserves the attention a dedicated PR & following discussions would have.

Reply:
I'd like to make sure it's clear that the diff you're referring to will not change behavior when using a global escape function replacement. Whether `&& escape === mustache.escape` is present or not affects the behavior of the newly introduced config object escape functions, and only config object escape functions. This particular diff is an effort to completely preserve previous behavior while allowing the useful new config object behavior to be more flexible than replacing the global escape function.

The abscence or presence of this check `&& escape === mustache.escape` should only have a bearing on the newly added functionality that this PR is focused on. If you're thinking of the more invasive changes suggested in https://github.com/pineapplemachine/mustache.js/compare/render-with-provided-escape-function...pineapplemachine:render-config-invasive, then I think you're right that making those changes a separate PR open for a separate discussion would be a good idea. If you're saying that this current PR should be broken up, then I don't think I'm on the same page. Unless I made an unintentional mistake, the previous behavior should be completely preserved, with the exception that sections will no longer discard a tags argument when rendering their contents, since I thought that was pretty clearly unexpected behavior. (https://github.com/janl/mustache.js/pull/764#discussion_r502277123)

Reply:
> This particular diff is an effort to completely preserve previous behavior while allowing the useful new config object behavior to be more flexible than replacing the global escape function.

Truly appreciated, impressive piece of work 👏 

Thanks for your point of view on what would be a good fit for a followup discussion & PR targeted to what you've raised related to escaping number values.